feat: Introduce activationThreshold/minMetricValue for Azure scalers

JorTurFer opened the pull request:
~~I have added the test case to every already existing e2e tests except in app-insights. They are quite complex and I need to review it with calm (or maybe wait till anyone migrates it)~~

All azure e2e test have been migrated to golang and updated to cover this new cases. Azure Monitor and Azure Event Hub haven't been done because there were already existing e2e test. 

In Azure Application Inisghts, I had to use goroutines to keep the value alive in order to query it from Azure service, I have tried several things but other options produce flaky tests, this seems more reliable. There were several configurations tested in typescript e2e test, but they are basically the same with different TriggerAuthentication configurations (secrets, envs, aad-pod), due to we don't have aad-pod-identity deployed in our clusters and secrets/envs are quite similar, I have migrated only the e2e that uses secrets

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))
- [x] Tests have been added

Fixes https://github.com/kedacore/keda/issues/3259
Fixes https://github.com/kedacore/keda/issues/3228
Related to #2800
Related to https://github.com/kedacore/keda-docs/pull/818

JorTurFer commented:
/run-e2e azure*

JorTurFer commented:
/run-e2e azure*

JorTurFer commented:
/run-e2e azure*

JorTurFer commented:
/run-e2e azure*

JorTurFer commented:
/run-e2e azure*

JorTurFer commented:
/run-e2e azure*

zroubalik commented on tests/scalers/azure-pipelines.test.ts:
```diff
@@ -69,13 +69,16 @@ test.serial('PoolID: Deployment should scale to 1 replica after queueing job', a
   var definitionID = parseInt(buildDefinitionID)
 
   await build.queueBuild(null, projectName, null, null, null, definitionID)
+  await sleep(60000);
```
Do you plan to do the migration to Go in this PR?

JorTurFer replied:
I didn't have, but I'll do it. This and application insights e2e tests

JorTurFer commented:
/run-e2e *pipelines*

JorTurFer commented:
/run-e2e azure*

JorTurFer commented:
/run-e2e azure*

JorTurFer commented:
/run-e2e azure*

JorTurFer commented:
I have migrated app-insights and az-pipelines e2e. Only Azure Monitor and Azure EventHub left because they didn't have e2e test, do you think that it's worth to add them from scratch in this PR?

JorTurFer commented:
[There is a spelling error related with e2e environment variables ](https://github.com/kedacore/keda/runs/7400213715?check_suite_focus=true#step:6:117). Once this PR is ready to merge, I'll update those variables to correct the error before merging (if I correct them now, e2e test will fail due to missing envs)

JorTurFer commented:
/run-e2e azure*

JorTurFer commented:
/run-e2e azure*

JorTurFer commented:
/run-e2e azure*

JorTurFer commented:
/run-e2e azure*

JorTurFer commented:
/run-e2e azure*

JorTurFer commented:
/run-e2e azure*

JorTurFer commented:
/run-e2e azure*

JorTurFer commented:
/run-e2e azure*

JorTurFer commented:
/run-e2e azure*

JorTurFer commented:
/run-e2e azure*

JorTurFer commented:
/run-e2e azure*

JorTurFer commented:
/run-e2e azure*

zroubalik commented:
> I have migrated app-insights and az-pipelines e2e. Only Azure Monitor and Azure EventHub left because they didn't have e2e test, do you think that it's worth to add them from scratch in this PR @zroubalik ?

I think we can keep it for another PR.

JorTurFer commented:
> I think we can keep it for another PR.

This PR is ready to review in that case

zroubalik commented on tests/scalers_go/azure_application_insights/azure_application_insights_test.go:
```diff
@@ -0,0 +1,244 @@
+	CreateKubernetesResources(t, kc, testNamespace, data, templates)
+
+	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, deploymentName, testNamespace, 0, 60, 1),
+		"replica count should be 0 after a minute")
```
```suggestion
		"replica count should be 0 after 1 minute")
```

zroubalik commented on tests/scalers_go/azure_pipelines/azure_pipelines_test.go:
```diff
@@ -0,0 +1,308 @@
+	t.Log("--- testing scale up ---")
+	queueBuild(t, connection)
+	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, deploymentName, testNamespace, maxReplicaCount, 60, 1),
+		"replica count should be 2 after a minute")
```
```suggestion
		"replica count should be 2 after 1 minute")
```

zroubalik commented on tests/scalers_go/azure_pipelines/azure_pipelines_test.go:
```diff
@@ -0,0 +1,308 @@
+func testScaleDown(t *testing.T, kc *kubernetes.Clientset) {
+	t.Log("--- testing scale down ---")
+	assert.True(t, WaitForPodCountInNamespace(t, kc, testNamespace, minReplicaCount, 60, 5),
+		"pod count should be 0 after a minute")
```
```suggestion
		"pod count should be 0 after 1 minute")
```

zroubalik reviewed (commented):
Looking good, just a minor nit, using `1` instead of `a` is imho more explicit and clear on a quick look, even though it has the same meaning.

zroubalik commented:
> Looking good, just a minor nit, using `1` instead of `a` is imho more explicit and clear on a quick look, even though it has the same meaning.

@JorTurFer could you please also rebase, I think I have fixed bunch of these `a` -> `1` in a previous PR.

JorTurFer commented:
/run-e2e azure*

JorTurFer commented:
> @JorTurFer could you please also rebase, I think I have fixed bunch of these `a` -> `1` in a previous PR.

done the rebase and updated the code :)
